premerge/post-submit-testing.md

```diff
@@ -0,0 +1,202 @@
+clusters. These Buildbot workers will use custom builders to launch a build
+wrapped in a kubernetes pod and report the results back to the buildbot master.
+When the job is finished, the pod will complete and capacity will be available
+for another build, or if there is nothing left to test GKE will see that there
+is nothing running on one of the nodes and downscale the node pool.
+
+### Annotated Builder
+
+llvm-zorg has multiple types of builders. We plan on using an AnnotatedBuilder.
+AnnotatedBuilders allow for the build to be driven using a custom python script
+rather than directly dictating the shell commands that should be run to perform
+the build. We need the flexibility of the AnnotatedBuilder to deploy jobs on the
+cluster. AnnotatedBuilder based builders also enable deploying changes without
+needing to restart the buildbot master. Without this, we have to wait for an
+administrator of the LLVM buildbot master to restart it before our changes get
+deployed. This could significantly delay updates or responses to incidents,
+especially before the system is fully stable.
+
+### Build Distribution
+
+We want to be able to take advantage of the autoscaling functionality of the new
+cluster to efficiently utilize resources. To do this, we plan on having the
+AnnotatedBuilder script launch builds as kubernetes pods. This allows for
+kubernetes to assign the builds to nodes and also allows autoscaling through the
+same mechanism that Github Actions Runner Controller (ARC) uses to autoscale.
+This enables us to quickly process builds at peak times and not pay for extra
+capacity when commit traffic is quiet, ensuring our resource use is efficient
+while still providing fast feedback.
+
+Using the kubernetes API inside of a python script (our AnnotatedBuilder
+implementation) to launch builds does add some complexity. However, we belive
+the additional complexity is justified as it allows us to achieve our goals
+while maintaining efficient resource usage.
+
+### Testing Configuration
+
+By testing configuration, we mean both the environment that the tests run in,
+and the set of tests that run. The testing configuration will be as close to the
+premerge configuration as possible. We will be running all tests inside the same
+container with the same scripts (the `monolithic-linux.sh` and
+`monolithic-windows.sh` scripts) used by the premerge testing. However, there
+will be one main difference between the premerge and postcommit testing
+configurations. In the postcommit configuration we propose testing all projects
+on every commit rather than only testing the projects that themselves changed or
+had dependencies that changed. We propose this for two main reasons. Firstly,
+Buildbot does not have support for heterogenous build configurations. This means
+that testing a different set of projects within a single builder depending upon
+the contents of the commit could easily cause problems. More notifications could
```

Comment:
What kind of problems?
From buildbot point of view, you would track the entire project (so no path-based-filtering), but the script could very well disable some tests based on the commit diff. Buildbot should be oblivious to this.